src/cmap/auth/mongodb_oidc.ts

```diff
@@ -36,32 +39,30 @@ export interface IdPServerResponse {
  * @public
  * @experimental
  */
-export interface OIDCCallbackContext {
+export interface OIDCResponse {
+  accessToken: string;
+  expiresInSeconds?: number;
   refreshToken?: string;
-  timeoutSeconds?: number;
-  timeoutContext?: AbortSignal;
-  version: number;
 }
 
 /**
  * @public
  * @experimental
  */
-export type OIDCRequestFunction = (
-  info: IdPServerInfo,
-  context: OIDCCallbackContext
-) => Promise<IdPServerResponse>;
+export interface OIDCCallbackParams {
+  timeoutContext: AbortSignal;
```

Comment:
> is it possible that someone could want to make a DB call on another MongoClient

Highly unlikely.  This callback is intended to communicate with an IDP which is unrelated to the driver.

Reply:
> What is the expectation around mishandling timeouts? what if the promise never resolves should we mitigate that risk in the driver?

I'd place this firmly in the "not the driver's problem" section.

> Not all things out there support signals or ever will support signals. For example is it possible that someone could want to make a DB call on another MongoClient?

I think the real argument you're making here is that the Node.js driver should support `AbortSignal` in the same way as applying CSOT 🙂 

I'd stick with a simple API and idiomatic API here (which this is) and not worry about these more complex use cases until there is a genuine demand for them.

Reply:
> the driver is responsible for enforcing timeouts, if possible.

Well, the driver can choose to ignore the result of the callback and treat it as failed if it detects the timeout expiring, but it _can't_ actually enforce anything, right?

> I also think CSOT makes the driver responsible for enforcing it 

Well, the spec says "This is to account for the human interaction required to complete the callback. In this case, the callback is not subject to CSOT".

Reply:
If you're using the same interface for machine and human callbacks, you'll need to make a distinction based on the auth mechanism parameter (`OIDC_CALLBACK` vs `OIDC_HUMAN_CALLBACK`).  For the machine callback, it *is* subject to CSOT if available, and it *is* expected to be cancelled if possible in the driver's language.   For Python we currently cannot cancel, only check the remaining time after any given operation.

Reply:
Talked about this a bit in a team meeting: Either we leave the responsibility of interruption entirely to the callback or we set up a Promise.race which will let the driver regain control but it won't stop the callback from pending.

In the former case, we can generally leave things be, pass the signal but we don't use it ourselves. If it hangs so be it, and 

In the latter case, I am not sure about the importance of the details of the outcome there but the user would be left with a callback that is going to, ideally, eventually fulfill but the value/error will not be handled by the driver. 

CSOT (relevant context: https://jira.mongodb.org/browse/DRIVERS-2347) should work somewhat like this, CSOT will be raced with processing waitQueueMembers for connection checkOut. If CSOT expires and the pool was creating a connection to service an operation, that creation is not interrupted but the operation does reject, the otherwise healthy connection should be put into the connection pool.  

The AbortSignal is starting to make more sense to me if our goal is to provide a mechanism that _both_ returns control to the driver and notifies the user's callback of early termination. I think it will be an important detail that the user's abort handling runs before the driver's (so either the driver ensures its listener runs last or we chain signals to make sure our abort comes after theirs).

Providing a plain timeout ms value _may_ also still work, as long as we call setTimeout after we've started the user's callback, if we both use the same value for ms and we call setTimeout first the driver timeout will win. 

> but it can't actually enforce anything, right?

Yes, I do not believe it is possible to agnostically preempt the callback. Unless we used some kind of thread solution but that is well out of scope here.

Reply:
I renamed to `callbackTimeoutMS` like the spec and pass that, updating to call `setTimeout` now.

Reply:
As discussed we've used the abort signal here with the same pattern we've done with CSOT using a Timeout.